components/Cards/src/MDCCard.m

```diff
@@ -0,0 +1,238 @@
+}
+
+- (MDCShadowElevation)shadowElevationForState:(UIControlState)state {
+  NSNumber *elevation = _shadowElevations[@(state)];
+  if (state != UIControlStateNormal && elevation == nil) {
+    elevation = _shadowElevations[@(UIControlStateNormal)];
+  }
+  if (elevation != nil) {
+    return (CGFloat)[elevation doubleValue];
+  }
+  return 0;
+}
+
+- (void)setShadowElevation:(MDCShadowElevation)shadowElevation forState:(UIControlState)state {
+  _shadowElevations[@(state)] = @(shadowElevation);
+
+  [self updateShadowElevation];
+}
+
+- (void)updateShadowElevation {
+  CGFloat elevation = [self shadowElevationForState:self.state];
+  if (!MDCCGFloatEqual(((MDCShadowLayer *)self.layer).elevation, elevation)) {
+    self.layer.shadowPath = [self boundingPath].CGPath;
+    [(MDCShadowLayer *)self.layer setElevation:elevation];
+  }
+}
+
+- (void)setBorderWidth:(CGFloat)borderWidth forState:(UIControlState)state {
+  _borderWidths[@(state)] = @(borderWidth);
+
+  [self updateBorderWidth];
+}
+
+- (void)updateBorderWidth {
+  CGFloat borderWidth = [self borderWidthForState:self.state];
+  self.layer.borderWidth = borderWidth;
+}
+
+- (CGFloat)borderWidthForState:(UIControlState)state {
+  NSNumber *borderWidth = _borderWidths[@(state)];
+  if (state != UIControlStateNormal && borderWidth == nil) {
+    borderWidth = _borderWidths[@(UIControlStateNormal)];
+  }
+  if (borderWidth != nil) {
+    return (CGFloat)[borderWidth doubleValue];
+  }
+  return 0;
+}
+
+- (void)setBorderColor:(UIColor *)borderColor forState:(UIControlState)state {
+  _borderColors[@(state)] = borderColor;
+
+  [self updateBorderColor];
+}
+
+- (void)updateBorderColor {
+  CGColorRef borderColorRef = [self borderColorForState:self.state].CGColor;
+  self.layer.borderColor = borderColorRef;
+}
+
+- (UIColor *)borderColorForState:(UIControlState)state {
+  UIColor *borderColor = _borderColors[@(state)];
+  if (state != UIControlStateNormal && borderColor == nil) {
+    borderColor = _borderColors[@(UIControlStateNormal)];
+  }
+  return borderColor;
+}
+
+- (void)setShadowColor:(UIColor *)shadowColor forState:(UIControlState)state {
+  _shadowColors[@(state)] = shadowColor;
+
+  [self updateShadowColor];
+}
+
+- (void)updateShadowColor {
+  CGColorRef shadowColor = [self shadowColorForState:self.state].CGColor;
+  self.layer.shadowColor = shadowColor;
+}
+
+- (UIColor *)shadowColorForState:(UIControlState)state {
+  UIColor *shadowColor = _shadowColors[@(state)];
+  if (state != UIControlStateNormal && shadowColor == nil) {
+    shadowColor = _shadowColors[@(UIControlStateNormal)];
+  }
+  if (shadowColor != nil) {
+    return shadowColor;
+  }
+  return [UIColor blackColor];
+}
+
+- (void)setHighlighted:(BOOL)highlighted {
+  [super setHighlighted: highlighted];
+  if (highlighted) {
+    /**
+     Note: setHighlighted might get more touches began than touches ended hence the call
+     hence the call to startTouchEndedAnimationAtPoint before.
```

Comment:
copy edit: 
setHighlighted: can get getting more calls with YES than NO when clicking rapidly. To guard against ink never going away and darkening our card we call startTouchEndedAnimationAtPoint:completion:.